### ISSUE TITLE
Enable clippy::integer_arithmetic check

### ISSUE THREAD

**Issue by akhi3030** [near/nearcore#8210]
In https://github.com/near/nearcore/issues/8145 we've added clippy as part of the CI.

Now we can enable [clippy::integer_arithmetic](https://rust-lang.github.io/rust-clippy/master/index.html#integer_arithmetic) to help with preventing bugs like https://github.com/near/nearcore/pull/8179.

This essentially bans usage of any arithmetic operations. See [this gist](https://gist.github.com/pugachAG/a76b8624462bdc9efe32c02fbcfbf1dc) for a small subset of nearcore violations of this issue.

Also see [this comment](https://github.com/near/nearcore/issues/8210#issuecomment-1380224548) for a compelling argument in favour of enabling this.

Progress:
* #8369
* #8370
* #8464

**Comment by pugachAG** [comment]
@akhi3030 @nagisa just to confirm: we want to have `-D clippy::integer_arithmetic` for the whole nearcore?

**Comment by akhi3030** [comment]
> @akhi3030 @nagisa just to confirm: we want to have `-D clippy::integer_arithmetic` for the whole nearcore?

Ideally we do.  But it also depends on how much work it will be to enable this.  I suspect it will be quite a bit of work.

**Comment by nagisa** [comment]
That’s correct. Protocol code is not immune to the issues caused by unchecked arithmetic, and the results are likely to be as disastrous. If making this lint denied is not a viable approach, we can always start with a `-W` or perhaps `-D` and sprinkle `#[warn(...)]`s at the current locations where this is a problem. That way new lint violations are not introduced and, hopefully, the old ones get progressively fixed too.

**Comment by nagisa** [comment]
Question about this issue in particular: is this an issue about `integer_arithmetic` in particular? We should reword the title as such. Or is it meant to be a tracking issue for ideas about what other lints we should look at enabling in the near future? I have some ideas here :)

**Comment by pugachAG** [comment]
I've updated this issue to specifically track `integer_arithmetic`. I suggest we use #8145 to discuss other lints we want to enable. @nagisa I'm looking forward to your proposals :)

Enabling `-D clippy::integer_arithmetic` would be pretty intrusive, I've added examples of the violations we currently have in nearcore to the description. I personally don't mind that, but I'm afraid some people might find it too restrictive. I guess it is worth discussing this with the whole Protocol team before starting the implementation.

**Comment by nagisa** [comment]
> Enabling -D clippy::integer_arithmetic would be pretty intrusive, I've added examples of the violations we currently have in nearcore to the description. I personally don't mind that, but I'm afraid some people might find it too restrictive.

I’m pretty sure we want `integer_arithmetic` across the codebase. Right now it is a code-smell for multiple reasons. Quite like unchecked indexing with `x[y]`, the plain integer arithmetic operations currently are checked and will panic if the results don't fit. This is a DoS hazard, no matter where in `nearcore` these occur. Quite unlike `x[y]`, people aren't as indoctrinated to treat regular arithmetics with the suspicion they deserve. So in the end we’re walking a thin rope where any slip-up means a `neard` DoS vulnerability, quite like the one described in [this postmortem](https://github.com/near/nearcore/discussions/8230).

Provided presence of bare integer arithmetic, it is important that the developers _prove_ for each and every one of them that the result will always fit for any possible inputs that operation might see. I’m willing to bet a token that such proof does not happen for the vast majority of the operations we have or add to nearcore. More than likely people go by a gut feel, or perhaps aren’t aware of the pitfalls associated with these operations in the first place.

An easy way out of this gotta-prove-and-double-check-everything conundrum is to make a call on what kind of handling of an overflow is appropriate for an operation. In some places saturation is appropriate, in others wrap-around is better, elsewhere a widening operation might be the best bet. Panic is almost(?) never a good option for `neard` and the lint in question makes sure that the engineers become aware of this.

---

Deconstructing one of the examples:

> !s.is_empty() && s.find('.') == Some(s.len() - 1)

is alright _right now_, but only because `s.is_empty()` condition was verified before the subtraction. There are plenty of innocuous-looking future modifications to this code that would be incorrect though: extracting `s.len() - 1` into a variable before the `s.is_empty()` check comes to mind as a particularly plausible one. If a check for some multi-byte character were added (like, if we made `AccountId`s fully unicode and supported full width `．` as a separator), a `s.len() - 3` might also show up in this check, which might then panic despite the `is_empty` check.

The alternative is `s.len().checked_sub(1)`. Besides no longer carrying a risk of a panic, it also verifies the `s.is_empty()` condition in the same operation! And is much more resilient to the future changes too.

Having to deal with such contortions may also suggest users to think about how to write the code in the way that describes the intent, rather than how to achieve a desired result (i.e. approach the problem more declaratively/functionally.) What this specific snippet of code does is checking whether there's exactly one `.` at the end of the (sub-)string. If the code was written to describe exactly that, it would be quite a bit easier to grok too.

**Comment by pugachAG** [comment]
I would like to propose implementing this in the following way:
1. Fix `integer_arithmetic` issues in core packages (`near-primitives`, `near-account-id`, etc.). I can take care of this, so I can get a better understanding of what is the proper fix for some of the common violations we have. 
2. Update [style.md](https://github.com/near/nearcore/blob/master/docs/practices/style.md) with the details regarding why  `integer_arithmetic` is discouraged and what are the common fixes (we can use some examples from step 1 for that).
3. Ask people in Protocol to fix `integer_arithmetic` the the packages they own.

@akhi3030 @nagisa please let me know if that makes sense to you.

**Comment by nagisa** [comment]
This approach seems okay to me. While we're fixing the occurrences, we can enable the lint on a per-crate basis for the time being too (with a `#![deny(clippy::integer_arithmetic)]` attribute at the crate root.) Let me know if you’d like me to write something for the style doc.

**Comment by pugachAG** [comment]
@nagisa good point regarding adding `deny` as we fix `integer_arithmetic` issues in the package, will use that.

>  Let me know if you’d like me to write something for the style doc.

That would be great, thanks! I was going to essentially copy&paste [your comment](https://github.com/near/nearcore/issues/8210#issuecomment-1380224548) there, so it would definitely be better if you write something yourself :)

**Comment by pugachAG** [comment]
@nagisa what would you suggest to use in case when arithmetic operation is actually safe? One example would be division and reminder operations for unsigned types. To me it seems like using `wrapping_div` or `saturating_div` would make it confusing, since reader could assume that overflow could actually happen in this case. One real-world example is [`get_num_seats_per_shard`](https://github.com/near/nearcore/blob/a56420c594f2230932c84ec856ea528b0812abf0/core/primitives/src/utils.rs#L427) function. I would prefer to use `.checked_div(..).unwrap()` in this case to emphasise that operation is safe by explicitly unwrapping, but I'm not sure if that is the "idiomatic" solution here.

**Comment by nagisa** [comment]
`checked_div().expect("can't fail")` is also a good option. But `unwrap` documenting that the operation can't fail is a good way to go about it. If it is performance-sensitive `unwrap_unchecked()` with a `SAFE:` comment explaining why and how this is unreachable.

**Comment by nagisa** [comment]
https://github.com/rust-lang/rust-clippy/issues/10200 is worth looking at.

**Comment by pugachAG** [comment]
@nagisa I've discovered an issue when using `derive` with `arbitrary::Arbitrary`, see  https://github.com/near/nearcore/pull/8408 for more details. Adding `#[allow(clippy::integer_arithmetic)]` on top of `derive` doesn't help. Do you have any suggestion on how we can deal with that?

**Comment by nagisa** [comment]
It is a bug in the implementation of the lint https://github.com/rust-lang/rust-clippy/issues/10185. Unfortunately I don't know if there's a way to address it other than waiting for it to get fixed (minimum 12 weeks) or possibly switching to the `arithmetic_side_effects` lint, which might not suffer from this issue.

**Comment by nagisa** [comment]
I guess I could look into fixing this in `arbitrary` too :)

**Comment by nagisa** [comment]
https://github.com/rust-fuzz/arbitrary/pull/140

**Comment by pugachAG** [comment]
> I guess I could look into fixing this in arbitrary too :)

Awesome, thanks for the quick fix! 

> or possibly switching to the `arithmetic_side_effects` lint, which might not suffer from this issue.

I've just checked and unfortunately that doesn't help:
```
error: arithmetic operation that can potentially result in unexpected side-effects
   --> core/primitives/src/types.rs:833:30
    |
833 | #[derive(Deserialize, Debug, arbitrary::Arbitrary)]
    |                              ^^^^^^^^^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arithmetic_side_effects
    = note: this error originates in the derive macro `arbitrary::Arbitrary` (in Nightly builds, run with -Z macro-backtrace for more info)
```
